src/mlpack/core/tree/vantage_point_tree/dual_tree_traverser_impl.hpp

```diff
@@ -0,0 +1,567 @@
+  {
+    // If the reference node contains a point we should calculate all
+    // base cases with this point.
+    if (referenceNode.IsFirstPointCentroid())
+    {
+      const size_t queryEnd = queryNode.Begin() + queryNode.Count();
+      for (size_t query = queryNode.Begin(); query < queryEnd; ++query)
+        rule.BaseCase(query, referenceNode.Point(0));
+      numBaseCases += queryNode.Count();
+    }
+    // We have to recurse down the reference node.  In this case the recursion
+    // order does matter.  Before recursing, though, we have to set the
+    // traversal information correctly.
+    double leftScore = rule.Score(queryNode, *referenceNode.Left());
+    typename RuleType::TraversalInfoType leftInfo = rule.TraversalInfo();
+    rule.TraversalInfo() = traversalInfo;
+    double rightScore = rule.Score(queryNode, *referenceNode.Right());
+    numScores += 2;
+
+    if (leftScore < rightScore)
+    {
+      // Recurse to the left.  Restore the left traversal info.  Store the right
+      // traversal info.
+      traversalInfo = rule.TraversalInfo();
+      rule.TraversalInfo() = leftInfo;
+      Traverse(queryNode, *referenceNode.Left());
+
+      // Is it still valid to recurse to the right?
+      rightScore = rule.Rescore(queryNode, *referenceNode.Right(), rightScore);
+
+      if (rightScore != DBL_MAX)
+      {
+        // Restore the right traversal info.
+        rule.TraversalInfo() = traversalInfo;
+        Traverse(queryNode, *referenceNode.Right());
+      }
+      else
+        ++numPrunes;
+    }
+    else if (rightScore < leftScore)
+    {
+      // Recurse to the right.
+      Traverse(queryNode, *referenceNode.Right());
+
+      // Is it still valid to recurse to the left?
+      leftScore = rule.Rescore(queryNode, *referenceNode.Left(), leftScore);
+
+      if (leftScore != DBL_MAX)
+      {
+        // Restore the left traversal info.
+        rule.TraversalInfo() = leftInfo;
+        Traverse(queryNode, *referenceNode.Left());
+      }
+      else
+        ++numPrunes;
+    }
+    else // leftScore is equal to rightScore.
+    {
+      if (leftScore == DBL_MAX)
+      {
+        numPrunes += 2;
+      }
+      else
+      {
+        // Choose the left first.  Restore the left traversal info.  Store the
+        // right traversal info.
+        traversalInfo = rule.TraversalInfo();
+        rule.TraversalInfo() = leftInfo;
+        Traverse(queryNode, *referenceNode.Left());
+
+        rightScore = rule.Rescore(queryNode, *referenceNode.Right(),
+            rightScore);
+
+        if (rightScore != DBL_MAX)
+        {
+          // Restore the right traversal info.
+          rule.TraversalInfo() = traversalInfo;
+          Traverse(queryNode, *referenceNode.Right());
+        }
+        else
+          ++numPrunes;
+      }
+    }
+  }
+  else
+  {
+    // If the reference node contains a point we should calculate all
+    // base cases with this point.
+    if (referenceNode.IsFirstPointCentroid())
```

Comment:
Ah hang on, but we could end up with duplicated base cases like this.  Let me think for a while if there might be a better way to avoid duplications but preserve this type of technique.  Maybe you are right that traversing with a query node and a reference point could work, but that would require a lot of refactoring of all the dual-tree algorithms so I'd prefer to avoid that...